### ISSUE TITLE
revalidatePath() not working as expected for Parallel Routes, or for Parallel + Intercepting Routes (modal)

### ISSUE THREAD

**Issue by magoz** [vercel/next.js#54173]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.6.0: Thu Jun  8 23:57:12 PDT 2023; root:xnu-8020.240.18.701.6~1/RELEASE_X86_64
    Binaries:
      Node: 20.5.1
      npm: 9.8.0
      Yarn: 1.22.19
      pnpm: 8.6.12
    Relevant Packages:
      next: 13.4.17
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: N/A
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/magoz/revalidatepath-parallel-routes-bug

### To Reproduce

Click on [Parallel+Intercepting Modal](https://revalidatepath-parallel-routes-bug.vercel.app/parallel-intercepting).

### Describe the Bug

When clicking on the `Trigger Server Action` button, the server action is triggered just by fetching a new random number and passing it to a client component.

It works perfectly for the [Home Page](https://revalidatepath-parallel-routes-bug.vercel.app/), and a [Leaf Page](https://revalidatepath-parallel-routes-bug.vercel.app/leaf).

But for a Page that uses Parallel + Intercepting Routes (modal), `revalidatePath('/parallel-intercepting')` seems to fetch the new data, and the component appears to receive the new props (see client console), but it doesn't re-render.

For a page that uses Parallel Routes (modal without Intercepting), works as expected on hard navigation, but on soft navigation we get the same issue than the Parallel + Intercepting Routes case.

### Expected Behavior

`revalidatePath()` to work as expected for Parallel Routes, and for Parallel + Intercepting Routes.

### Which browser are you using? (if relevant)

Chrome 115.0.5790.170

### How are you deploying your application? (if relevant)

pnpm dev

<sub>[NEXT-1706](https://linear.app/vercel/issue/NEXT-1706/revalidatepath-not-working-as-expected-for-parallel-routes-or-for)</sub>

**Comment by kght6123** [comment]
I too was plagued by this bug.
I temporarily added the following settings to work around it, but the cache is no longer meaningful.
```
export const dynamic = "force-dynamic";
```

**Comment by adelegard** [comment]
@kght6123 that doesn't seem to fix it for me. Tried with @magoz example too - and doesn't work in there either.

**Comment by grzegorzpokorski** [comment]
Hi, I also have a problem with `revalidatePath` invoked in server action which is executed on the page which is available in intercepted and pararel route. I also execute server action with `revalidatePath` inside as a calback function in `useTransition` hook. In folowing reproduction [link](https://github.com/grzegorzpokorski/next-usetransition-server-actions-revalidation-issue) you can notice that value of flag that tells you whether there is a pending transition is once set to true, but when data is successfully fetched inside server action this flag from `useTransition` does not inform about it - insted inform that this function is still pending.
Invoking `revalidatePath` in server action in pararell + intercepted route cause "demage" of navigation - when you want to change route or go backward of your browsing history to close intercepted "modal", intercepted route does not hide etc. Please check [this reproduction of bugs](https://github.com/grzegorzpokorski/next-usetransition-server-actions-revalidation-issue). which I notice.

**Comment by grzegorzpokorski** [comment]
> Hi, I also have a problem with `revalidatePath` invoked in server action which is executed on the page which is available in intercepted and pararel route. I also execute server action with `revalidatePath` inside 'startTransition' function from `useTransition` hook. In folowing reproduction [link](https://github.com/grzegorzpokorski/next-usetransition-server-actions-revalidation-issue) you can notice that value of flag that tells you whether there is a pending transition is once set to true, but when data is successfully fetched inside server action this flag from `useTransition` does not inform about it - insted inform that this function is still pending. Invoking `revalidatePath` in server action in pararell + intercepted route cause "demage" of navigation - when you want to change route or go backward of your browsing history to close intercepted "modal", intercepted route does not hide etc. Please check [this reproduction of bugs](https://github.com/grzegorzpokorski/next-usetransition-server-actions-revalidation-issue). which I notice.

It also occurs when in server action I use `cookies()` from `next/headers`

**Comment by adelegard** [comment]
I also find that this same problem exists for `revalidateTag`.

**Comment by grzegorzpokorski** [comment]
https://github.com/vercel/next.js/issues/55911

**Comment by kght6123** [comment]
@adelegard sorry. It certainly doesn't seem to be working.

**Comment by ats1999** [comment]
revalidation is breaking in all way!

**Comment by adelegard** [comment]
Any update on this? This bug continues to cause us pain and having to do weird hacky things to work around it.

**Comment by tinleym** [comment]
Given that parallel routes are the recommended way to render modals (https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#modals), this bug breaks the recommended and documented pattern for submitting forms in modals.

**Comment by adelegard** [comment]
> Given that parallel routes are the recommended way to render modals (https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#modals), this bug breaks the recommended way to submit forms in modals.

Totally. I don't understand how the community isn't freaking out about this tbh.

**Comment by master4o** [comment]
seems like a main issue now, cuz block using forms in modals

**Comment by Xananax** [comment]
Is there any workaround to this?

`export const dynamic = "force-dynamic";` does not work in server components.

I am not really going to bring anything new, but I'll explain my setup below, in case someone else ends up here and wants to know if their bug is similar to the one explained here.

<details>

<summary>My setup</summary>

- `auth/[slug]/page.tsx` loads the logged in user, then loads login/signup pages
- `@modal/(.)auth/[slug]/page.tsx` does the same in a modal

This seems to be the setup recommended in the docs. It breaks in weird ways. 

At first, everything seems to work as intended. I am receiving the form errors I send back from my action:
![image](https://github.com/vercel/next.js/assets/836082/86566ca0-ee12-4ae7-b38c-7441f2f718d9)

They do get displayed
![image](https://github.com/vercel/next.js/assets/836082/2f04d7fb-b0be-423e-a9f9-5186e68eb20f)

I have a `console.log` that reads the formState result, and it shows up:
![image](https://github.com/vercel/next.js/assets/836082/85cc3f7f-029d-4805-85ce-e71f704fde8c)

This is all working as expected

----

However if the form submits (data is validated), then the form stops updating.

I do receive the proper response from the server:
![image](https://github.com/vercel/next.js/assets/836082/d1338281-5ef2-4a8f-b0be-f0b4e7b4b106)

But the form stops updating, no console.log shows up.

It seems the form is incapable of updating after these lines in my server action:
```js
  const { error } = await supabase.auth.signUp({
    email,
    password,
    options: {
      emailRedirectTo,
    },
  });
  ```

If I `return` anything before those lines, everything works; anything after those lines, nothing happens.

I'm using Supabase SSR, and it modifies cookies. 

Manually adding a `revalidatePath("/auth/signup")` didn't change anything

</details>

--------------------

It really looks like the recommended way doesn't work for user authentication, despite that being both an example in the docs, and the main motivation to use parallel routes to my eyes.

If anyone has a workaround, I'd be happy to know.

**Comment by master4o** [comment]
> Is there any workaround to this?
> 
> `export const dynamic = "force-dynamic";` does not work in server components.
> 
> I am not really going to bring anything new, but I'll explain my setup below, in case someone else ends up here and wants to know if their bug is similar to the one explained here.
> 
> My setup
> * `auth/[slug]/page.tsx` loads the logged in user, then loads login/signup pages
> * `@modal/(.)auth/[slug]/page.tsx` does the same in a modal
> 
> This seems to be the setup recommended in the docs. It breaks in weird ways.
> 
> At first, everything seems to work as intended. I am receiving the form errors I send back from my action: ![image](https://user-images.githubusercontent.com/836082/285273855-86566ca0-ee12-4ae7-b38c-7441f2f718d9.png)
> 
> They do get displayed ![image](https://user-images.githubusercontent.com/836082/285273955-2f04d7fb-b0be-423e-a9f9-5186e68eb20f.png)
> 
> I have a `console.log` that reads the formState result, and it shows up: ![image](https://user-images.githubusercontent.com/836082/285274133-85cc3f7f-029d-4805-85ce-e71f704fde8c.png)
> 
> This is all working as expected
> 
> However if the form submits (data is validated), then the form stops updating.
> 
> I do receive the proper response from the server: ![image](https://user-images.githubusercontent.com/836082/285274506-d1338281-5ef2-4a8f-b0be-f0b4e7b4b106.png)
> 
> But the form stops updating, no console.log shows up.
> 
> It seems the form is incapable of updating after these lines in my server action:
> 
> ```js
>   const { error } = await supabase.auth.signUp({
>     email,
>     password,
>     options: {
>       emailRedirectTo,
>     },
>   });
> ```
> 
> If I `return` anything before those lines, everything works; anything after those lines, nothing happens.
> 
> I'm using Supabase SSR, and it modifies cookies.
> 
> Manually adding a `revalidatePath("/auth/signup")` didn't change anything
> 
> It really looks like the recommended way doesn't work for user authentication, despite that being both an example in the docs, and the main motivation to use parallel routes to my eyes.
> 
> If anyone has a workaround, I'd be happy to know.

I used bad, but working idea.

1. implement useFormState.
2. In Server Action return object with message: 'ok', if response status.ok
3.  if (loginState?.message === "ok" && sentCode) {
      setAuthCookie({ token: loginState.token });
      in modal component if mesage in state from useFormState === 'ok', invoke second separated server action only with cookies logic.
      
 It works for me, but it is very dirty solution. Need to wait issue resolve :(

**Comment by Xananax** [comment]
Oh wow, that's pretty smart! But yea, it doesn't feel good.

**Comment by dclark27** [comment]
Also need a fix to this! Our forms in parallel routes are breaking our navigation.

https://github.com/vercel/next.js/assets/9562283/f743b5ff-43de-42e2-9de3-fd052501f814

related: https://github.com/vercel/next.js/issues/58201
related: https://github.com/vercel/next.js/issues/58180

**Comment by tsnieman** [comment]
I'm experiencing this bug and it presents pretty much exactly like @dclark27's video above.

This is a pretty huge bug with some very key features of nextjs — who can we tag to look into this?

**Comment by tsnieman** [comment]
Here's as minimal reproduction as I could make:
🐛 https://github.com/tsnieman/nextjs-parallel-route-pain/

PLEASE let me know if there's more that i can provide. this bug is a **tremendous pain** that **my team doesn't have a workaround for**.

and here's a video of the bug in action:

https://github.com/vercel/next.js/assets/595711/a98bb2ca-ab4e-48f9-bc66-1d6641cc5ce9

**Comment by esemczak** [comment]
This issue absolutely persists and can be easily reproduced. Can we have at least acknowledgement from the team that they are aware and working on it? Parallel routes are completely useless in current state.

**Comment by numman-ali** [comment]
There is a commit that went in recently by @ztanner (thank you!) which mentions the nextUrl in the header can cause issues with server actions 
https://github.com/vercel/next.js/pull/59175

Perhaps a solution in the same area could be found for revalidateTag and and revalidatePath? @ztanner do you have any thoughts?

**Comment by feedthejim** [comment]
we're definitely investigating :)

**Comment by numman-ali** [comment]
Hey @ztanner thanks for putting the latest commit in, really appreciated 🙏 
I've pulled the latest canary that contains your changes but still doesn't seem to working. 
I've tested with router.refresh client side and a revalidateTag/Path on a server action and both times the 404 page is called 
Any thoughts on what might be happening?

**Comment by feedthejim** [comment]
@numman-ali do you have a reproduction?

**Comment by sinafath** [comment]
yeah the issue still exists in 14.0.5-canary.18

**Comment by bosheski** [comment]
Issue still exists in 14.0.5-canary.22 @feedthejim

**Comment by jeremy-code** [comment]
@feedthejim

I am not certain whether this is the same issue @bosheski, @sinafath, and @numman-ali are experiencing, but this is my minimal reproduction of an issue with intercepting routes on `14.0.5-canary.29`: [jeremy-code/next-intercepting-bug](https://github.com/jeremy-code/next-intercepting-bug). 

![next-intercepting-bug](https://github.com/vercel/next.js/assets/43259194/f1c61b9d-d4c1-467e-86a4-974980fef237)

The server action calls `revalidatePath("/")`. When there is no intercepting route, running the server action from the default page works as intended. When there is an intercepting route, running the server action from either the modal or the default page renders the 404 page.

Note that removing app/[slug]/page.tsx results in the same behavior, I included it to show that the 404 page it returns is not the result of the server rendering the default page at the non-intercepted route. 

I imagine the expected behavior would be for the current context's data to be reloaded, but not the route, or that the modal remains rendered rather than the non-intercepted page.

Let me know if there's anything I need to clarify.

**Comment by stefan-joseph** [comment]
The most recent update fixed the issue on the default page for the form, however, the bug still persists when submitting the form from a modal using parallel routing.

**Comment by jasonbert** [comment]
I think this behaviour may be related to the issues description here? I've got a modal displaying a blog post coming from a paralell/intercepted route, which has a "favourite" (heart) toggle button that calls a server action and revalidates the cache by tag. When "closing" the modal (`router.back()`), the route changes but doesn't remove the modal.

1) Trigger modal/paralell/intercepted route by clicking on blog link (`next/link`).
![image](https://github.com/vercel/next.js/assets/2347572/45a3ea50-44ab-48c1-9902-676f4ae14ae9)

2) Click on "favourite" toggle button, triggers server action and revalidates by tag.
![image](https://github.com/vercel/next.js/assets/2347572/cda2932c-eb38-4dcd-9bf5-ab2b8642cf52)

3) Click on "close" (X) button to close the modal (`router.back()`). **The route has changed, but the modal hasn't been removed.**
![image](https://github.com/vercel/next.js/assets/2347572/670a2b91-9689-4757-b56d-c408a426be88)


Modal Component
```typescript
'use client'

import { useRouter } from "next/navigation";
import React, { MouseEventHandler, PropsWithChildren, useCallback, useEffect, useRef } from 'react';
import { Heading2 } from '../heading';
import styles from './index.module.scss';

interface ModalProps extends PropsWithChildren {
    header?: string | React.ReactNode
    footer?: React.ReactNode
    onClose?: () => void
}

export default function Modal({ children, header, footer, onClose }: ModalProps) {
    const overlay = useRef(null);

    const handleClose = useCallback(() => {
        if (onClose) onClose();
    }, [onClose]);

    const handleClick: MouseEventHandler = useCallback((e) => {
        if (onClose && e.target === overlay.current) {
            onClose();
        }
    }, [onClose]);

    const handleKeyDown = useCallback((e: KeyboardEvent) => {
        if (onClose && e.key === "Escape") onClose();
    }, [onClose]);

    useEffect(() => {
        document.addEventListener("keydown", handleKeyDown);
        return () => document.removeEventListener("keydown", handleKeyDown);
    }, [handleKeyDown]);

    return (
        <div ref={overlay} className={styles.Modal} onClick={handleClick}>
            <div className={styles.Modal__container}>
                <div className={styles.Modal__header}>
                    {header &&
                        <div className={styles.Modal__headerContent}>
                            {typeof header === 'string' ?
                                <Heading2>{header}</Heading2> :
                                header
                            }
                        </div>
                    }
                    <div className={styles.Modal__headerClose}>
                        <button onClick={handleClose} className={styles.Modal__headerCloseButton}>X</button>
                    </div>
                </div>
                <div className={styles.Modal__content}>
                    {children}
                </div>
                {footer &&
                    <div className={styles.Modal__footer}>
                        <div className={styles.Modal__footerContent}>
                            {footer}
                        </div>
                    </div>
                }
            </div>
        </div>
    )
}

export function RouterModal(props: ModalProps) {
    const router = useRouter();
    const handleClose = useCallback(() => {
        if (props.onClose) {
            props.onClose();
        }

        router.back();
    }, [router]);

    return <Modal {...props} onClose={handleClose} />
}
```

Action
```typescript
'use server'

export async function toggleBlogFavoriteAction(blogId: string) {
    await toggleBlogFavorite(blogId);

    revalidateTag(BLOG_FAVORITES_CACHE_TAG);
}
```

Component
```typescript
'use client';

import { startTransition, useOptimistic } from "react";
import { PrimaryButton } from "~f/framework/button";
import { toggleBlogFavoriteAction } from "./blogAction";

export default function BlogFavoriteToggleButton({ blogId, isFavorite }: { blogId: string, isFavorite: boolean }) {
    const [optimisticIsFavorite, setOptimisticIsFavorite] = useOptimistic(isFavorite);

    async function toggle() {
        startTransition(() => {
            setOptimisticIsFavorite(!optimisticIsFavorite);
        });

        await toggleBlogFavoriteAction(blogId);
    };

    return (
        <PrimaryButton onClick={toggle} style={{ minWidth: 50 }}>{optimisticIsFavorite ? '❤️' : '🖤'}</PrimaryButton>
    )
}
```

Source: https://github.com/scorado-ltd/scorado-examples-nextjs-cosmos

**Comment by jeremy-code** [comment]
I have discovered where the 404 page is coming from that I mentioned in my previous comment. It looks like Next.js is trying to load the default component (in my case, `app/default.tsx`) at the root. If there is any function at all, it now loads the modal (but not `app/page.tsx`). If `default.tsx` does load a component, it will load that in addition to the modal. The server action continues to run like before, but without any errors. I have updated my reproduction repository to demonstrate what I mean at [jeremy-code/next-intercepting-bug](https://github.com/jeremy-code/next-intercepting-bug).

Here is a gif of what I am referring to:
![CleanShot 2024-01-03 at 14 22 43](https://github.com/vercel/next.js/assets/43259194/91decb16-39e2-4d2c-af4c-1a749042d94f)
